### PR TITLE
feat: t4 dataset metadata + closed-loop reproducer/metrics refactor

### PR DESCRIPTION
**Pull request by xtk8532704**
## What this PR does

Propagates T4 dataset IDs through bag metadata, and refactors closed-loop reproducer metrics / stuck handling to better match Autoware parity (yaw-gated frame picking, stuck = repeat ∧ ego nearly stopped, arc-length teleport, strong-brake metric, unified `*_steps` / `*_count` / `*_segments` naming).

### T4 dataset metadata (`bag_metadata`)

Reads/writes `t4_dataset_id` and `t4_dataset_version_id` from rosbag metadata JSON so converted datasets can carry the upstream T4 dataset identity.

### Closed-loop summary metrics (`summary.json`)

Aggregated by `scenario_generation/closed_loop_eval.aggregate` from per-segment rows.

#### Naming convention (event families)

For `{m}` ∈ `{collision, near_miss, strong_brake}`:

| Concept | Key |
| --- | --- |
| steps in state | `{m}_steps` |
| discrete events (rising edge, 3-frame falling-edge debounce) | `{m}_count` |
| segments with ≥1 such step | `{m}_segments` |
| fraction of steps | `{m}_step_rate` |
| fraction of segments | `{m}_segment_rate` |

#### 1. Run / denominators

| Key | Meaning |
| --- | --- |
| `total_steps` | Σ per-segment steps run |
| `n_segments` | number of segments / routes scored |
| `n_routes` | routes rolled out this run |
| `near_miss_thresh` | clearance threshold used for near-miss (m) |
| `strong_brake_mps2` | tangential accel threshold for strong brake (default `-3.0`) |
| `elapsed_sec` | wall time for the eval run |

#### 2. Safety events

| Family | Definition | Keys |
| --- | --- | --- |
| **collision** | OBB ego↔neighbor overlap | `collision_steps`, `collision_count`, `collision_segments`, `collision_step_rate`, `collision_segment_rate` |
| **near-miss** | finite clearance ≤ `near_miss_thresh` | `near_miss_steps`, `near_miss_count`, `near_miss_segments`, `near_miss_step_rate`, `near_miss_segment_rate` |
| **strong brake** | realized tangential accel ≤ `strong_brake_mps2` | `strong_brake_steps`, `strong_brake_count`, `strong_brake_segments`, `strong_brake_step_rate`, `strong_brake_segment_rate` |

#### 3. Clearance

| Key | Meaning |
| --- | --- |
| `global_min_clearance` | min over finite segment min-clearances (m) |
| `mean_segment_min_clearance` | mean of finite segment min-clearances (m) |
| `mean_segment_mean_clearance` | mean of finite segment mean-clearances (m) |

#### 4. Reproducer / unstick telemetry

| Key | Meaning |
| --- | --- |
| `snap_count` | Σ hard GT-pose teleports (renamed from `total_snaps`) |
| `expand_count` | Σ stage-1 search-radius widen events |
| `teleport_count` | Σ stage-2 arc-length teleport events |
| `normal_steps` | Σ cursor steps in `normal` (advancing) |
| `repeat_steps` | Σ cursor steps in `repeat` (republishing same frame) |

#### 5. Termination

| Key | Meaning |
| --- | --- |
| `terminated_counts` | `{reason: n_segments}` histogram (`max_steps` / `goal` / `stuck`, …) |

**Breaking:** old keys (`n_collision_steps`, `n_near_miss_steps`, `n_segments_with_*`, `total_*`) are gone; consumers must use the new names.

### Reproducer / unstick (Autoware-aligned)

- **Yaw gate** (`DEFAULT_YAW_GATE_RAD = π/2`): xy-only KDTree candidates whose recorded heading differs from live ego by more than the gate are dropped, so U-turn / self-crossing routes do not snap to opposite-heading frames. If every candidate is wrong-heading, the cursor stays in `repeat` rather than picking a bad frame.
- **Cursor state**: exposes only `normal` / `repeat` (plus run-length + cumulative step counts). Widen / teleport stay rollout-owned.
- **Stuck definition**: stuck only when `repeat` **and** ego speed ≤ `STUCK_SPEED_MPS` (0.5). A stopped ego whose reproducer is still advancing does not escalate.
- **Teleport target**: `RouteTimeline.index_ahead_by_arc_length` (bag arc length ahead), mirroring Autoware `find_perturb_index_along_bag`.

## Testing

- Extended `scenario_generation/tests/test_reproducer_unstick.py`: stuck requires repeat, expand/teleport counts, yaw-gate rejects opposite-heading candidates on a self-crossing route, arc-length teleport target.
- Call-site key renames covered in `test_r2lpl_lifelong_replay.py` / mining tooling.
- C++ bag metadata unit test for the new T4 fields.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: rlvr/autoresearch/tools/mine_collisions_reproducer.py [hunk 1]
```diff
@@ -412,8 +412,8 @@ def _flush(buf_units, buf_keys):
     print("top hits (collisions desc, clearance asc):")
     for r in hits[:10]:
         print(
-            f"  {r['route']} {r['segment']}  collisions={r['n_collision_steps']:3d}  "
```
Comment (Owen-Liuyuxuan) [inline rlvr/autoresearch/tools/mine_collisions_reproducer.py:415]: @xtk8532704 It seems the renaming will affect quite many other PRs

**Review by go-sakayori — COMMENTED** [review]
Personally I think it is better to split the PR into 2: one for T4 dataset related and another for planning metrics.

File: scenario_generation/reproducer_rollout.py [hunk 2]
```diff
@@ -865,8 +894,9 @@ def _advance_step(s: _SegState, pred: np.ndarray, idx, device, timers, override=
                 s.last_turn_indicator = int(s.turn_hist[-1])
                 s.last_collision_uuid = None  # teleported -> next contact is a fresh collision
                 s.in_episode = False
-                s.prev_max_idx = s.cursor.max_idx_reached
+                s.prev_max_idx = cur.max_idx_reached
                 s.ego_stuck = 0
+                s.teleport_count += 1
                 s.n_snaps += 1
```
Comment (go-sakayori) [inline scenario_generation/reproducer_rollout.py:900]: whats the difference between teleport_count and snaps count

File: scenario_generation/perception_reproducer.py [hunk 3]
```diff
@@ -107,10 +146,12 @@ def step(self, sim_xy: np.ndarray, sim_speed: float, sim_time: float) -> int:
             )
             if self.search_radius <= 0.0:
                 # Degenerate mode: nearest recorded frame, but still never rewind
-                # once this cursor has reached a later recorded frame.
+                # once this cursor has reached a later recorded frame. Always advances,
+                # so it is never in the ``repeat`` state.
                 idx = max(self.tl.nearest(sim_xy), self.max_idx_reached)
                 self._last_idx = idx
                 self.max_idx_reached = max(self.max_idx_reached, idx)
+                self._update_base_state(repeat=False)
```
Comment (danielsanchezaran) [inline scenario_generation/perception_reproducer.py:154]: Blocking: zero-radius mode does not necessarily advance. If the ego remains at the same location, `nearest()` can return the same frame on every tick, but this hard-codes `repeat=False`. That makes `normal_steps`/`repeat_steps` incorrect and, because unstick now requires `last_was_repeat`, prevents a stalled `search_radius=0` rollout from ever escalating. Please distinguish the initial selection from subsequent republishes of the same index and add a zero-radius regression test.

File: scenario_generation/reproducer_rollout.py [hunk 4]
```diff
@@ -710,7 +727,7 @@ def _pre_step(s: _SegState, gpu_transform: bool = False):
         s.prev_max_idx = idx
         s.stuck = 0
     else:
-        idx = s.cursor.step(s.live_pose[:2], s.dyn.speed, s.sim_time)
+        idx = s.cursor.step(s.live_pose[:2], s.dyn.speed, s.sim_time, sim_yaw=float(s.live_pose[2]))
```
Comment (danielsanchezaran) [inline scenario_generation/reproducer_rollout.py:730]: The `clock` branch above bypasses `cursor.step()` and never updates the new state counters. Clock progression is the R2LPL default, so those runs currently finish with `normal_steps=0` and `repeat_steps=0` regardless of `n_steps_run`. Please account for each clock tick as normal advancement (or explicitly expose separate clock telemetry) and test that the state-step totals match the executed steps in both progression modes.

File: diffusion_planner/valid_predictor_closed_loop.py [hunk 5]
```diff
@@ -136,6 +142,7 @@ def _eval_knobs(args: argparse.Namespace) -> dict:
         draw_every=args.draw_every,
         neighbor_history_mode="recorded",
         tracker_mode="perfect",
+        strong_brake_mps2=args.strong_brake_mps2,
```
Comment (danielsanchezaran) [inline diffusion_planner/valid_predictor_closed_loop.py:145]: This forwards `strong_brake_mps2` to each worker, but the multi-GPU parent calls `_merge_shards(out_dir, npz_root, near_miss_thresh)` and writes `summary.json` without the threshold. The sequential summary includes it, so the same metric has different persisted schemas and sharded strong-brake counts cannot be interpreted from the artifact alone. Please pass the threshold into the parent merge and add a sharded-summary test. Also consider doing final serialization after parent-owned fields such as `elapsed_sec` are added; today that field is added only after `_merge_shards` has already written the file.

**Review by danielsanchezaran — CHANGES_REQUESTED** [review]
Blocking on two state/telemetry gaps and the sharded summary contract. The focused reproducer/R2LPL tests pass (160 passed, 1 skipped), but none exercises these paths.

File: scenario_generation/closed_loop_eval.py [hunk 6]
```diff
@@ -90,28 +97,40 @@ def aggregate(rows: list[dict], near_miss_thresh: float) -> dict:
...
+        {
+            "global_min_clearance": float(min(finite_min_cl)) if finite_min_cl else float("inf"),
+            "mean_segment_min_clearance": float(np.mean(finite_min_cl))
+            if finite_min_cl
+            else float("inf"),
+            "mean_segment_mean_clearance": float(np.mean(finite_mean_cl))
+            if finite_mean_cl
+            else float("inf"),
+            # Reproducer stuck telemetry (Autoware parity) + snap count.
+            "snap_count": _sum("n_snaps"),
```
Comment (go-sakayori) [inline scenario_generation/closed_loop_eval.py:125]: Are you going to use `n_snaps` or `snap_count`?

**Review by danielsanchezaran — COMMENTED** [review]
Additional findings from a fresh pass over the reproducer/metrics logic — complementing the earlier change-request, not restating the zero-radius / clock-counter / sharded-summary points.

### Blocking — unstick escalation is dead in clock-replay mode (the R2LPL default)

`reproducer_rollout.py` `_advance_step` now gates stuck accumulation on `cur.last_was_repeat`:

```python
elif not cur.last_was_repeat:
    s.ego_stuck = 0
else:
    s.ego_stuck += 1
```

In the `clock` branch of `_pre_step`, `cursor.step()` is never called, so `last_was_repeat` never leaves its `reset()` default of `False`. So in clock mode `ego_stuck` is cleared every tick and **stage-1 widen / stage-2 teleport never fire**. The previous speed-only gate (`if s.dyn.speed >= 0.5 … else s.ego_stuck += 1`) *did* teleport a stalled ego in clock mode. `run_lifelong_r2lpl_rounds.py` defaults `timeline_progress_mode="clock"` and forwards `unstick_after=300`, so this silently disables the unstick safety-escalation on the default R2LPL path: a model that predicts a stop now stalls until the max-steps cap instead of being teleported past it. Same root cause as the clock/counter comment, but a functional regression rather than a telemetry gap — flagging separately.

### Medium — yaw gate is default-ON with no opt-out

`DEFAULT_YAW_GATE_RAD = π/2`, and `_seed_state` builds the cursor without plumbing `yaw_gate` while `_pre_step` always passes `sim_yaw`. Every rollout (mining, closed-loop eval, R2LPL) now applies the 90° heading gate with no flag to disable it through `_seed_state`/`run_segments_batched`. On tight junctions / roundabouts / self-crossing routes a forward in-radius candidate >90° off live heading is now dropped, so frame selection — and therefore previously-mined datasets/metrics — changes with no reproducibility escape hatch. Reasonable default; please expose a disable path.

### Medium — `aggregate` `.get(key, 0)` silently zeroes pre-rename rows

`closed_loop_eval.aggregate` reads the new names via `r.get("collision_steps", 0)` etc. Re-aggregating any pre-PR `segments*.jsonl` (old names `n_collision_steps` / `n_near_miss_steps` / …) now reports `collision/near_miss/strong_brake = 0` instead of raising `KeyError` like the old `r[...]` indexing did. The docstring frames `.get` as tolerating "older rows [that] may lack keys," but old rows *have* the data under different names — `.get` masks the rename into a silent under-report on cross-version re-aggregation. Consider raising when a row carries old keys but not new.

### Nits

- `route_timeline.index_ahead_by_arc_length` returns `start_idx` unchanged for `distance_m <= 0`; the teleport then `reset()`s onto the same anchor and can re-stall/re-teleport indefinitely. Non-default (`unstick_advance_m=5.0`), but the old code always advanced ≥1 frame.
- `--strong_brake_mps2` has no sign validation despite the "negative" help text; a positive value makes `accels <= thresh` fire nearly every step.
- `strong_brake_mps2` isn't plumbed through `run_segments_batched` (only `render_segment`), so the batched mining path always uses `-3.0`. Fine today; an inconsistency if a batched caller ever exposes the knob.
- `test_bag_metadata.cpp::ValidJsonAllFields` asserts the t4 fields are empty despite its name (the populated path is covered by `T4DatasetFieldsAreLoaded`) — naming only.

### Verified NOT issues

`accels[]`/`clearances[]` index alignment (incl. step 0 and warmup); strong-brake sign/units (`dyn.accel` is longitudinal, `<= -3.0` correctly detects hard decel); `_event_count` rising-edge + 3-frame debounce edge cases; `near_mask` equivalence to the old `valid_cl <= thresh`; `accels is None` guard states. The T4 `bag_metadata` C++ (safe missing-field handling, struct field order, tests) is correct, and the breaking key-rename is fully covered across in-repo consumers (`train.py`, `valid_predictor_closed_loop.py`, `mine_collisions_reproducer.py`, `run_lifelong_r2lpl_rounds.py`, tests) with no un-updated readers in code/docs/configs.

File: scenario_generation/reproducer_rollout.py [hunk 7]
```diff
@@ -816,41 +833,53 @@ def _advance_step(s: _SegState, pred: np.ndarray, idx, device, timers, override=
...
         if s.unstick_after > 0:
-            if s.dyn.speed >= 0.5:
+            cur = s.cursor
+            if s.dyn.speed > STUCK_SPEED_MPS:
                 # Moving again: clear the counter and undo any temporary cursor widening so
                 # frame selection returns to the nominal search_radius.
                 s.ego_stuck = 0
-                s.cursor.restore_radius()
+                cur.restore_radius()
+            elif not cur.last_was_repeat:
```
Comment (danielsanchezaran) [inline scenario_generation/reproducer_rollout.py:857]: **Blocking (functional regression, not just the counter gap noted elsewhere).** This stuck gate keys off `cur.last_was_repeat`, but the `clock` branch of `_pre_step` never calls `cursor.step()`, so `last_was_repeat` stays at its `reset()` default of `False`. In clock mode `ego_stuck` is therefore cleared every tick and **stage-1 widen / stage-2 teleport never fire**. The previous speed-only gate (`if s.dyn.speed >= 0.5 … else s.ego_stuck += 1`) did teleport a stalled ego in clock mode. Since `run_lifelong_r2lpl_rounds.py` defaults `timeline_progress_mode="clock"` and forwards `unstick_after=300`, this silently disables the unstick safety-escalation on the default R2LPL path — a model that predicts a stop now stalls to the max-steps cap instead of being teleported past it. Please drive the stuck counter (or an equivalent) in clock mode too, and add a clock-mode teleport regression test.

File: scenario_generation/perception_reproducer.py [hunk 8]
```diff
@@ -40,6 +40,17 @@
 _SPEED_GAP_RATIO = 2.0
 _SPEED_GAP_MIN_REC = 3.0  # m/s
 
+# Heading gate (rad): a candidate recorded frame is only eligible if its recorded ego
+# yaw is within +/- this of the live ego yaw. Stops the xy-only KDTree from grabbing a
+# future opposite-heading frame (U-turn / self-crossing route) that would re-center the
+# scene rotated ~180 deg. 90 deg is lenient enough for normal curves/junctions.
+DEFAULT_YAW_GATE_RAD = np.pi / 2.0
```
Comment (danielsanchezaran) [inline scenario_generation/perception_reproducer.py:47]: Medium: this yaw gate is default-ON and there's no way to turn it off in practice. `_seed_state` builds the cursor as `PerceptionReproducer(tl, search_radius=…, timers=…)` without plumbing `yaw_gate`, and `_pre_step` now always passes `sim_yaw`, so every rollout (mining, closed-loop eval, R2LPL) applies the 90° gate. On tight junctions / roundabouts / self-crossing routes a forward in-radius candidate >90° off live heading is now dropped, so frame selection — and previously-mined datasets/metrics — changes with no reproducibility escape hatch. Reasonable default, but please plumb `yaw_gate` through `_seed_state`/`run_segments_batched` so it can be disabled.

File: scenario_generation/closed_loop_eval.py [hunk 9]
```diff
@@ -72,15 +72,22 @@ def resolve_npz_roots(npz_root) -> list[Path]:
...
     n_seg = len(rows)
     total_steps = sum(r["n_steps_run"] for r in rows)
-    total_collision_steps = sum(r["n_collision_steps"] for r in rows)
-    total_near_miss_steps = sum(r["n_near_miss_steps"] for r in rows)
-    total_snaps = sum(r["n_snaps"] for r in rows)
 
-    n_seg_collision = sum(1 for r in rows if r["n_collision_steps"] > 0)
-    n_seg_near_miss = sum(1 for r in rows if r["n_near_miss_steps"] > 0)
+    def _sum(key: str) -> int:
+        return sum(int(r.get(key, 0)) for r in rows)
```
Comment (danielsanchezaran) [inline scenario_generation/closed_loop_eval.py:87]: Medium: `.get(key, 0)` on the *new* names silently zeroes pre-rename rows. Re-aggregating any pre-PR `segments*.jsonl` (old names `n_collision_steps`/`n_near_miss_steps`/…) now reports `collision/near_miss/strong_brake = 0` instead of raising `KeyError` like the old `r[...]` indexing. The docstring frames `.get` as tolerating rows that "lack keys," but old rows *have* the data under different names — this masks the rename into a silent under-report on cross-version re-aggregation. Suggest raising when a row carries the old keys but not the new.

File: scenario_generation/route_timeline.py [hunk 10]
```diff
@@ -287,6 +287,27 @@ def nearest(self, xy: np.ndarray) -> int:
...
+
+        Accumulates the 2D distance between consecutive recorded ego poses forward from
+        ``start_idx`` and returns the first index whose cumulative arc length reaches
+        ``distance_m`` (clamped to the last frame if the route ends first). Arc length along
+        the recorded path, not Euclidean distance from a query point. Used by the rollout's
+        unstick teleport to pick a target ahead; mirrors Autoware ``find_perturb_index_along_bag``.
+        """
+        n = len(self)
+        start_idx = int(max(0, min(start_idx, n - 1)))
+        if distance_m <= 0.0 or start_idx + 1 >= n:
```
Comment (danielsanchezaran) [inline scenario_generation/route_timeline.py:302]: Nit: returning `start_idx` unchanged for `distance_m <= 0` means the teleport (`reset(tgt)` with `tgt == max_idx_reached`) lands back on the same anchor with no forward progress, so a stalled ego can re-stall and re-teleport to the same frame indefinitely. Non-default (`unstick_advance_m=5.0`), but the old teleport always advanced ≥1 frame — worth guaranteeing at least `start_idx+1` here.

File: diffusion_planner/valid_predictor_closed_loop.py [hunk 11]
```diff
@@ -58,6 +58,12 @@ def parse_args() -> argparse.Namespace:
     # --- tunable knobs (default to the closed-loop mining config) ---
     p.add_argument("--device", type=str, default="cuda", help="'cuda' or 'cpu'")
     p.add_argument("--near_miss_thresh", type=float, default=0.5, help="near-miss clearance (m)")
+    p.add_argument(
+        "--strong_brake_mps2",
```
Comment (danielsanchezaran) [inline diffusion_planner/valid_predictor_closed_loop.py:62]: Nit: no sign validation despite the "negative" help text — a positive `--strong_brake_mps2` makes `accels <= thresh` fire on nearly every step. Also `strong_brake_mps2` is only plumbed through `render_segment`, not `run_segments_batched`, so the batched mining path always uses the `-3.0` default (fine today, but an inconsistency if a batched caller ever exposes the knob).

**Review by danielsanchezaran — COMMENTED** [review]
Switching to inline comments on the specific lines (supersedes my earlier top-level note). These complement the existing change-request and don't restate the zero-radius / clock-counter / sharded-summary points.

Verified NOT issues, so they aren't chased: `accels[]`/`clearances[]` index alignment (incl. step 0), strong-brake sign/units (`dyn.accel` is longitudinal, `<= -3.0` is correct hard-decel), `_event_count` rising-edge + 3-frame debounce, `near_mask` equivalence, and `accels is None` guard states. The T4 `bag_metadata` C++ is correct and the breaking key-rename is fully covered across in-repo consumers. One naming nit: `test_bag_metadata.cpp::ValidJsonAllFields` asserts the t4 fields are empty despite its name (populated path is covered by `T4DatasetFieldsAreLoaded`).

**Comment by xtk8532704** [comment]
Closing this PR: it mixed T4 bag metadata + reproducer unstick + metrics rename, and has known blocking issues (zero-radius repeat, clock-mode unstick, sharded summary).

Splitting into focused PRs off `tier4-main`, using `kem/cl-rb-redlight-metrics` as the source of truth:
- PR1: T4 bag metadata (`kem/refactor-reproducer-test-PR1`)
- PR2: Autoware-aligned reproducer unstick (`kem/refactor-reproducer-test-PR2`)
- PR3: Nested closed-loop metrics (`kem/refactor-reproducer-test-PR3`)

Will link the new PRs here once opened.

**Comment by xtk8532704** [comment]
Split PRs (from `kem/cl-rb-redlight-metrics`):
- PR1 T4 bag metadata: https://github.com/tier4/Diffusion-Planner/pull/278
- PR2 Unstick: https://github.com/tier4/Diffusion-Planner/pull/279
- PR3 Nested metrics (stacked on PR2): (opening now)

**Comment by xtk8532704** [comment]
Split complete:
- https://github.com/tier4/Diffusion-Planner/pull/278 — T4 bag metadata
- https://github.com/tier4/Diffusion-Planner/pull/279 — Autoware-aligned unstick
- https://github.com/tier4/Diffusion-Planner/pull/280 — Nested closed-loop metrics (stacked on #279)
